File: Java/src/Act4/Estudiante.java
package Act4;

import java.util.HashMap;
import java.util.Map;

class Estudiante {
    private String nombre;
    private int id;
    private Map<Curso, Double> calificaciones;

    public Estudiante(String nombre, int id) {
        this.nombre = nombre;
        this.id = id;
        this.calificaciones = new HashMap<>();
    }

    public String getNombre() {
        return nombre;
    }

    public int getId() {
        return id;
    }

    public void registrarCurso(Curso curso) {
        if (!calificaciones.containsKey(curso)) {
            calificaciones.put(curso, null);
        }
    }

    public void agregarCalificacion(Curso curso, double calificacion) {
        if (calificaciones.containsKey(curso)) {
            calificaciones.put(curso, calificacion);
        } else {
            System.out.println("El estudiante no está registrado en este curso.");
        }
    }

    public double calcularPromedio() {
        double suma = 0.0;
        int contador = 0;
        for (Double calificacion : calificaciones.values()) {
            if (calificacion != null) {
                suma += calificacion;
                contador++;
            }
        }
        if(contador < 0){
            return 0;
        }else{
            return  suma / contador;
        }

    }

    @Override
    public String toString() {
        return "Estudiante{" +
                "nombre='" + nombre + '\'' +
                ", id=" + id +
                ", promedio=" + calcularPromedio() +
                '}';
    }
}

File: KotlinCode/src/Act4/Estudiante.kt
package Act4

import java.util.HashMap

internal class Estudiante(var nombre: String?,
                          var id: Int,
                          var calificaciones: MutableMap<Curso?, Double?>?) {

    fun registrarCurso(curso: Curso?) {
        calificaciones?.containsKey(curso)?.let {
            if (!it) {
                calificaciones?.put(curso, null)
            }
        }
    }

    fun agregarCalificacion(curso: Curso?, calificacion: Double) {
        if (calificaciones?.containsKey(curso) == true) {
            calificaciones?.put(curso, calificacion)
        } else {
            println("El estudiante no está registrado en este curso.")
        }
    }

    fun calcularPromedio(): Double {
        var suma = 0.0
        var contador = 0
        for (calificacion in calificaciones?.values!!) {
            if (calificacion != null) {
                suma += calificacion
                contador++
            }
        }
        if (contador < 0) {
            return 0.0
        } else {
            return suma / contador
        }
    }

    override fun toString(): String {
        return "Estudiante{" +
                "nombre='" + nombre + '\'' +
                ", id=" + id +
                ", promedio=" + calcularPromedio() +
                '}'
    }
}
